R/with-qenv.R

```diff
@@ -0,0 +1,411 @@
+    "  Use `quenv$<variable>` or `quenv[[\"<variable>\"]]`to access variables.",
+    sep = "\n"
+  )
+
+  # contents/bindings # nolint
+  var_names <- ls(x)
+  var_names_hidden <- setdiff(ls(x, all.names = TRUE), var_names)
+
+  if (length(var_names) + length(var_names_hidden) > 0L) {
+    var_classes <- sapply(var_names, function(vn) toString(class(get(vn, envir = x, inherits = FALSE))))
+    var_classes_hidden <- sapply(var_names_hidden, function(vn) toString(class(get(vn, envir = x, inherits = FALSE))))
+
+
+    longest_name <- max(nchar(c(var_names, var_names_hidden)))
+    longest_class <- max(nchar(c(var_classes, var_classes_hidden)))
+
+    contents <- if (is.finite(longest_name)) {
+      paste(
+        sprintf(sprintf("  $ %%-0%is", longest_name), var_names),
+        sprintf(sprintf("%%-0%is", longest_class), var_classes),
+        vapply(var_names, function(v) .object_info(get(v, envir = x)), character(1)),
+        sep = " : ", collapse = "\n"
+      )
+    }
+
+    contents_hidden <- paste(
+      sprintf(sprintf("  $ %%-0%is", longest_name), var_names_hidden),
+      sprintf(sprintf("%%-0%is", longest_class), var_classes_hidden),
+      vapply(var_names_hidden, function(v) .object_info(get(v, envir = x)), character(1)),
+      sep = " : ", collapse = "\n"
+    )
+
+    contents_all <- c(
+      if (!identical(contents, "")) sprintf("bindings:\n%s", contents),
+      if (!identical(contents_hidden, "")) sprintf("hidden bindings:\n%s", contents_hidden)
+    )
+  } else {
+    contents_all <- "This quenv is empty."
+  }
+
+  # code
+  code <- attr(x, "code")
+  code <-
+    if (identical(code, list())) {
+      ""
+    } else {
+      expressions <- vapply(code, function(x) paste(sprintf("      %s", deparse(x)), collapse = "\n"), character(1L))
+      paste(
+        "  {",
+        paste(expressions, collapse = "\n"),
+        "  }",
+        sep = "\n"
+      )
+    }
+  code <- if (!identical(code, "")) sprintf("code:\n%s", code)
+
+  # conditions
+  conditions <- lapply(get_conditions(x, "all"), sprintf, fmt = "  %s")
+  conditions <- unlist(
+    mapply(
+      function(value, name) sprintf("%s:\n%s", name, paste(value, collapse = "\n")),
+      value = conditions, name = names(conditions)
+    )
+  )
+
+  # closing message
+  footer <- sprintf("parent: %s", format(parent.env(x)))
+
+  c(
+    header,
+    "",
+    contents_all,
+    "",
+    code,
+    "",
+    conditions,
+    "",
+    footer
+  )
+}
+
+
+#' @rdname quenv
+#' @export
+#' @keywords internal
+print.quenv <- function(x, ...) {
+  cat(format(x, ...), sep = "\n")
+}
+
+
+#' @export
+#' @keywords internal
+`[.quenv` <- function(x, ...) { # nolint
+  stop("Use `quenv$<variable>` or `quenv[[\"<variable>\"]]`to access variables.")
+}
+
+
+#' @export
+#' @keywords internal
+`$<-.quenv` <- function(x, name, value) { # nolint
+  stop(
+    "Direct assignment is forbidden as it cannot be tracked. ",
+    "Use `with( <quenv>, { <name> <- <value> })` instead."
+  )
+}
```

Comment:
instead of `$<-.quenv`, `[[<-.qenv` it is possible to just `lockEnvironment` and unlock only inside `within`